### DISCUSSION TITLE
Is it possible to have an SQL reference with syntax diagrams?

### DISCUSSION THREAD

**Discussion by ramazanpolat** [ClickHouse/ClickHouse#33158] [Q&A] [answered]
Since the **substitions for table and database names** feature is introduced in v21.12 release, I've been thinking that if the SQL syntax of ClickHouse has a formal grammar that can be expressed with WSN or not.

If it is, can we have a syntax diagram like this:

![image](https://user-images.githubusercontent.com/3527695/147374643-3383d5c2-031e-4e34-a387-d9ffcea30e2b.png)

Which looks much better than this:
```sql
[WITH expr_list|(subquery)]
SELECT [DISTINCT [ON (column1, column2, ...)]] expr_list
[FROM [db.]table | (subquery) | table_function] [FINAL]
[SAMPLE sample_coeff]
[ARRAY JOIN ...]
[GLOBAL] [ANY|ALL|ASOF] [INNER|LEFT|RIGHT|FULL|CROSS] [OUTER|SEMI|ANTI] JOIN (subquery)|table (ON <expr_list>)|(USING <column_list>)
[PREWHERE expr]
[WHERE expr]
[GROUP BY expr_list] [WITH ROLLUP|WITH CUBE] [WITH TOTALS]
[HAVING expr]
[ORDER BY expr_list] [WITH FILL] [FROM expr] [TO expr] [STEP expr]
[LIMIT [offset_value, ]n BY columns]
[LIMIT [n, ]m] [WITH TIES]
[SETTINGS ...]
[UNION  ...]
[INTO OUTFILE filename [COMPRESSION type] ]
[FORMAT format]
```

A better example would be the following [IBM product ducumentation](https://www.ibm.com/docs/en/db2-for-zos/12?topic=statements-alter-table) page which even lets users to click on expressions to see their production rules and relevant documentation in the same page.

![image](https://user-images.githubusercontent.com/3527695/147374776-2c875812-56a8-4a81-8df4-b19948ed471c.png)

**Comment by alexey-milovidov** [comment, marked as answer]
Currently no. Unfortunately our parser is not declarative and we cannot transform it to this graph.
I want to address this problem by rewriting it. Previous attempt with ANTLR was unsuccessful.

You can find incomplete, unsupported ANTLR grammar definition for ClickHouse here:
https://github.com/ClickHouse/ClickHouse/tree/master/utils/antlr

It covers maybe 80% of the language.
